File: project-data/ingest.js
import { pipeline } from '@huggingface/transformers'
import { readdir, readFile } from 'fs/promises'
import path from 'path'
import { fileURLToPath } from 'url'

const QDRANT_URL = 'http://localhost:6333'
const COLLECTION = 'project-docs'
const VECTOR_SIZE = 1024
const BATCH_SIZE = 50
const MIN_CHUNK_LENGTH = 50

const __dirname = path.dirname(fileURLToPath(import.meta.url))

function getCategory(filePath) {
  if (filePath.includes('/adrs/') || filePath.includes('\\adrs\\')) return 'adr'
  if (filePath.includes('/incidents/') || filePath.includes('\\incidents\\')) return 'incident'
  if (filePath.includes('/runbooks/') || filePath.includes('\\runbooks\\')) return 'runbook'
  if (filePath.includes('/api/') || filePath.includes('\\api\\')) return 'api'
  if (filePath.includes('/pages/') || filePath.includes('\\pages\\')) return 'page'
  if (filePath.includes('/features/') || filePath.includes('\\features\\')) return 'feature'
  return 'core'
}

function chunkMarkdown(text) {
  const lines = text.split('\n')
  const chunks = []
  let title = ''
  let currentSection = ''
  let currentLines = []

  const flush = () => {
    const content = currentLines.join('\n').trim()
    if (content.length >= MIN_CHUNK_LENGTH) {
      chunks.push({ section: currentSection, text: content })
    }
    currentLines = []
  }

  for (const line of lines) {
    if (line.startsWith('# ')) {
      title = line.replace(/^#\s+/, '').trim()
      currentLines.push(line)
    } else if (line.match(/^#{2,3}\s/)) {
      flush()
      currentSection = line.replace(/^#{2,3}\s+/, '').trim()
      currentLines = [line]
    } else {
      currentLines.push(line)
    }
  }
  flush()

  return { title, chunks }
}

async function walkDir(dir) {
  const files = []
  const entries = await readdir(dir, { withFileTypes: true })
  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name)
    if (entry.isDirectory()) {
      files.push(...(await walkDir(fullPath)))
    } else if (entry.name.endsWith('.md')) {
      files.push(fullPath)
    }
  }
  return files
}

async function ensureCollection() {
  const res = await fetch(`${QDRANT_URL}/collections/${COLLECTION}`)
  if (res.status === 200) {
    console.log(`Collection "${COLLECTION}" already exists — recreating`)
    await fetch(`${QDRANT_URL}/collections/${COLLECTION}`, { method: 'DELETE' })
  }
  await fetch(`${QDRANT_URL}/collections/${COLLECTION}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      vectors: { size: VECTOR_SIZE, distance: 'Cosine' },
    }),
  })
  console.log(`Collection "${COLLECTION}" created`)
}

async function upsertBatch(points) {
  const res = await fetch(`${QDRANT_URL}/collections/${COLLECTION}/points`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ points }),
  })
  if (!res.ok) {
    const body = await res.text()
    throw new Error(`Qdrant upsert failed: ${body}`)
  }
}

async function main() {
  console.log('Loading BAAI/bge-m3 model (first run downloads ~1.1 GB)...')
  const embedder = await pipeline('feature-extraction', 'BAAI/bge-m3')
  console.log('Model ready')

  await ensureCollection()

  const files = await walkDir(__dirname)
  console.log(`Found ${files.length} markdown files`)

  const points = []
  let pointId = 0

  for (const file of files) {
    const relativePath = path.relative(__dirname, file)
    const content = await readFile(file, 'utf-8')
    const category = getCategory(file)
    const { title, chunks } = chunkMarkdown(content)

    console.log(`  ${relativePath} → ${chunks.length} chunks`)

    for (let i = 0; i < chunks.length; i++) {
      const { section, text } = chunks[i]
      const output = await embedder(text, { pooling: 'mean', normalize: true })
      const vector = Array.from(output.data)

      points.push({
        id: pointId++,
        vector,
        payload: {
          text,
          source: relativePath,
          category,
          title: title || path.basename(file, '.md'),
          section,
          chunk_index: i,
        },
      })
    }
  }

  console.log(`\nUpserting ${points.length} points in batches of ${BATCH_SIZE}...`)
  for (let i = 0; i < points.length; i += BATCH_SIZE) {
    const batch = points.slice(i, i + BATCH_SIZE)
    await upsertBatch(batch)
    console.log(`  ${Math.min(i + BATCH_SIZE, points.length)}/${points.length} done`)
  }

  console.log('\nIngestion complete.')
}

main().catch((err) => {
  console.error(err)
  process.exit(1)
})

File: project-data/ingest.py
#!/usr/bin/env python3
"""
Ingest project-data markdown files into Qdrant using BAAI/bge-m3.
Run: python3 ingest.py
"""
import re
from pathlib import Path
from qdrant_client import QdrantClient
from qdrant_client.models import Distance, VectorParams, PointStruct
from sentence_transformers import SentenceTransformer

QDRANT_URL = "http://localhost:6333"
COLLECTION = "project-docs"
VECTOR_SIZE = 1024
BATCH_SIZE = 50
MIN_CHUNK_LENGTH = 50


def get_category(file_path: str) -> str:
    if '/adrs/' in file_path:      return 'adr'
    if '/incidents/' in file_path: return 'incident'
    if '/runbooks/' in file_path:  return 'runbook'
    if '/api/' in file_path:       return 'api'
    if '/pages/' in file_path:     return 'page'
    if '/features/' in file_path:  return 'feature'
    return 'core'


def chunk_markdown(text: str):
    lines = text.split('\n')
    title = ''
    current_section = ''
    current_lines = []
    chunks = []

    def flush():
        content = '\n'.join(current_lines).strip()
        if len(content) >= MIN_CHUNK_LENGTH:
            chunks.append({'section': current_section, 'text': content})
        current_lines.clear()

    for line in lines:
        if re.match(r'^# ', line):
            title = re.sub(r'^# ', '', line).strip()
            current_lines.append(line)
        elif re.match(r'^#{2,3} ', line):
            flush()
            current_section = re.sub(r'^#{2,3} ', '', line).strip()
            current_lines.append(line)
        else:
            current_lines.append(line)
    flush()

    return title, chunks


def main():
    base_dir = Path(__file__).parent

    print("Loading BAAI/bge-m3 model (first run downloads ~1.1 GB)...")
    model = SentenceTransformer('BAAI/bge-m3')
    print("Model ready\n")

    client = QdrantClient(url=QDRANT_URL)

    existing = [c.name for c in client.get_collections().collections]
    if COLLECTION in existing:
        print(f'Collection "{COLLECTION}" exists — recreating')
        client.delete_collection(COLLECTION)

    client.create_collection(
        collection_name=COLLECTION,
        vectors_config=VectorParams(size=VECTOR_SIZE, distance=Distance.COSINE),
    )
    print(f'Collection "{COLLECTION}" created\n')

    files = sorted(
        f for f in base_dir.rglob('*.md')
        if 'node_modules' not in f.parts
    )
    print(f"Found {len(files)} markdown files\n")

    points = []
    point_id = 0

    for file in files:
        relative = str(file.relative_to(base_dir))
        content = file.read_text(encoding='utf-8')
        category = get_category(str(file))
        title, chunks = chunk_markdown(content)

        print(f"  {relative} → {len(chunks)} chunks")

        for i, chunk in enumerate(chunks):
            text = chunk['text']
            vector = model.encode(text, normalize_embeddings=True).tolist()

            points.append(PointStruct(
                id=point_id,
                vector=vector,
                payload={
                    'text': text,
                    'source': relative,
                    'category': category,
                    'title': title or file.stem,
                    'section': chunk['section'],
                    'chunk_index': i,
                }
            ))
            point_id += 1

    print(f"\nUpserting {len(points)} points in batches of {BATCH_SIZE}...")
    for i in range(0, len(points), BATCH_SIZE):
        batch = points[i:i + BATCH_SIZE]
        client.upsert(collection_name=COLLECTION, points=batch)
        print(f"  {min(i + BATCH_SIZE, len(points))}/{len(points)} done")

    print(f"\nIngestion complete — {len(points)} points in collection '{COLLECTION}'")


if __name__ == '__main__':
    main()
